src/transformation/visitors/optional-chaining.ts

```diff
@@ -0,0 +1,187 @@
+import * as ts from "typescript";
+import * as lua from "../../LuaAST";
+import { TransformationContext } from "../context";
+import { assert, assertNever } from "../../utils";
+import { transformInPrecedingStatementScope } from "../utils/preceding-statements";
+import { transformPropertyAccessExpressionWithCapture, transformElementAccessExpressionWithCapture } from "./access";
+import { shouldMoveToTemp } from "./expression-list";
+
+function flattenChain(chain: ts.OptionalChain) {
+    assert(!ts.isNonNullChain(chain));
+    const links: ts.OptionalChain[] = [chain];
+    while (!chain.questionDotToken && !ts.isTaggedTemplateExpression(chain)) {
+        const nextLink: ts.Expression = chain.expression;
+        assert(ts.isOptionalChain(nextLink) && !ts.isNonNullChain(nextLink));
+        chain = nextLink;
+        links.unshift(chain);
+    }
+    return { expression: chain.expression, chain: links };
+}
+
+export interface ExpressionWithThisValue {
+    expression: lua.Expression;
+    thisValue?: lua.Expression;
+    isNewTemp?: boolean;
+}
+
+function transformExpressionWithThisValueCapture(
+    context: TransformationContext,
+    node: ts.Expression
+): ExpressionWithThisValue {
+    if (ts.isParenthesizedExpression(node)) {
+        return transformExpressionWithThisValueCapture(context, node.expression);
+    }
+    if (ts.isPropertyAccessExpression(node)) {
+        return transformPropertyAccessExpressionWithCapture(context, node, true);
```

Comment:
The main issue is that optional chaining cannot be transformed "normally", as the result of the optional chain will in the end be inside an if statement. If `transformCallExpression` were to handle this logic, it needs to somehow be able to tell the nodes higher in the AST (the rest of the optional chain) to put the result in the if statement.
The way I handled it here is largely based on `transformES2020` in typescript, which includes passing flags around. However, transformES2020 assumes all call expressions can take thisArg, which is not the case in TSTL, hence the state.

In hindsight, it might be possible to abandon `transformES2020`'s model and instead handle most of the logic within `transformCallExpression`, if some state were added in `TransformationContext` instead. This might be a better solution; I might try implementing this within the next few days and open another pull request.